### PR TITLE
[d3-sankey] Update to Version 0.5

### PR DESCRIPTION
**Pull request by tomwanzek**
* [BREAKING CHANGE] `SankeyLayout` API rewritten
* [BREAKING CHANGE] `SankeyLinkPathGenerator` removed. `sankeyLinkHorizontal` added as replacement.
* [CHORE] Update version numbers.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/d3/d3-sankey/releases/tag/v0.5.0>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by tomwanzek** [comment]
@gustavderdrache When you have a few minutes... Given that d3-sankey is still pre-1.0, the API was rewritten quite a bit. No major rush. Personally I like the latest changes, but (so far) the installed base of this D3 module is quite limited given its specificity relative to the other modules.

**Review by gustavderdrache — COMMENTED** [review]
A couple of typos in the docs but otherwise seems solid.

File: types/d3-sankey/index.d.ts [hunk 1]
```diff
@@ -129,133 +138,202 @@ export interface SankeyLinkMinimal<N extends SankeyExtraProperties, L extends Sa
...
-     * Set the sankey generator's links to the specified array of objects and returns this sankey layout generator.
+     * Set the Sankey generator's links accessor to a function returning the specified array of objects and returns this Sankey layout generator.
      *
      * @param links Array of links.
      */
     links(links: Array<SankeyLink<N, L>>): this;
+    /**
+     * Set the Sankey generator's links accessor to the specified function and returns this Sankey layout generator.
+     *
+     * @param nodes A links accessor function. The function is invoked when the Sankey layout is generated, being passed any arguments passed to the Sankey generator.
```
Comment (tomwanzek) [inline types/d3-sankey/index.d.ts]: Correct.

File: types/d3-sankey/index.d.ts [hunk 2]
```diff
@@ -264,4 +342,4 @@ export function sankey(): SankeyLayout<{}, {}>;
  * Sankey layout generator. These properties are IN EXCESS to the properties explicitly identified in the
  * SankeyLinkMinimal interface.
  */
-export function sankey<N extends SankeyExtraProperties, L extends SankeyExtraProperties>(): SankeyLayout<N, L>;
+export function sankeyLinkHorizontal<N extends SankeyExtraProperties, L extends SankeyExtraProperties>(): Link<any, SankeyLink<N, L>, [number, number]>;
```
Comment (tomwanzek) [inline types/d3-sankey/index.d.ts]: Am I assuming correctly that you are referring to the `sankeyLinkHorizontal` signature? I thought about it in the context of `sankeyLinkHorizontal` being syntactic sugar for fully configuring your own horizontal  link using **d3-shape's** `linkHorizontal()`.

Given that it is already preconfigured with `source` and `target` accessors and assumes the default `x` and `y` accessors, it did not seem necessary to introduce a generic for the `this` context. As the preconfigured accessors do not access a specific `this` context. By the same token, I fixed the third generic of the returned `Link` to `[number, number]`. In principle, we could add additional generics/signatures, but it seems we would move away from the intended syntactic sugar, to a full horizontal link as **d3-shape** provides it.

Over to you, I will already push a commit addressing the above two comments in the meantime.

Comment (tomwanzek) [inline types/d3-sankey/index.d.ts]: @gustavderdrache What do you think in light of my above comment? Should we broaden the signatures to reflect the full scope of generics supported by `Link<...>`? I do not have a strong preference, both approaches have pros and cons and are equally workable.

Comment (gustavderdrache) [inline types/d3-sankey/index.d.ts]: Nah, I think you have a good point. Carry on :)

**Review by gustavderdrache — APPROVED** [review]

**Comment by tomwanzek** [comment]
@gustavderdrache thanks.
@andy-ms Ready to merge. Thanks in advance.

**Pull request merged**